docs/architecture/deep-dives/state/index.md

```diff
@@ -0,0 +1,379 @@
+
+```typescript
+type ExpectedGlobalState = { myGlobalData: string };
+
+type ExpectedAccountState = { myUserData: string };
+
+const MY_GLOBAL_KEY_DEFINITION: KeyDefinitionLike = {
+  stateDefinition: { name: "myState" },
+  key: "myGlobalKey",
+};
+const MY_USER_KEY_DEFINITION: KeyDefinitionLike = {
+  stateDefinition: { name: "myState" },
+  key: "myUserKey",
+};
+
+export class MoveToStateProvider extends Migrator<10, 11> {
+  async migrate(migrationHelper: MigrationHelper): Promise<void> {
+    const existingGlobalData = await migrationHelper.get<ExpectedGlobalState>("global");
+
+    await migrationHelper.setGlobal(MY_GLOBAL_KEY_DEFINITION, {
+      myGlobalData: existingGlobalData.myGlobalData,
+    });
+
+    const updateAccount = async (userId: string, account: ExpectedAccountState) => {
+      await migrationHelper.setUser(MY_USER_KEY_DEFINITION, {
+        myUserData: account.myUserData,
+      });
+    };
+
+    const accounts = await migrationHelper.getAccounts<ExpectedAccountState>();
+
+    await Promise.all(accounts.map(({ userId, account }) => updateAccount(userId, account)));
+  }
+}
+```
+
+:::note
+
+`getAccounts` only gets data from the legacy account object that was used in `StateService`. As data
+gets migrated off of that account object the response from `getAccounts`, which returns a record
+where the key will be a users id and the value being the legacy account object.
+
+_TODO: Implement method just for user ids_
+
+:::
+
+### Example PRs
+
+_TODO: Include PR's_
+
+## Testing
+
+Testing business logic with data and observables can sometimes be cumbersome, to help make that a
+little easier, there are a suite of helpful "fakes" that can be used instead of traditional "mocks".
+Now instead of calling `mock<StateProvider>()` into your service you can instead use
+`new FakeStateProvider()`.
+
+_TODO: Refine user story_
+
+## Advanced usage
+
+### `update`
+
+The update method has options defined as follows:
+
+```typescript
+{ActiveUser|SingleUser|Global}State<T> {
+  // ... rest of type left out for brevity
+  update<TCombine>(updateState: (state: T, dependency: TCombine) => T, options?: StateUpdateOptions);
+}
+
+type StateUpdateOptions = {
+  shouldUpdate?: (state: T, dependency: TCombine) => boolean;
+  combineLatestWith?: Observable<TCombine>;
+  msTimeout?: number
+}
+```
+
+The `shouldUpdate` option can be useful to help avoid an unnecessary update, and therefore avoid an
+unnecessary emission of `state$`. You might want to use this to avoid setting state to `null` when
+it is already `null`. The `shouldUpdate` method gives you in it's first parameter the value of state
+before any change has been made to it and the dependency you have, optionally, provided through
+`combineLatestWith`. To avoid setting `null` twice you could call `update` like below:
+
+```typescript
+await myUserState.update(() => null, { shouldUpdate: (state) => state != null });
```

Comment:
❓ Is this accurate? Based on the description I'd expect the code to never update state once it's null. That is, if I manage to set `state === null`, then `shouldUpdate` would always return false, and thus cancel **every** update.

Reply:
Yep, good call. I Reworded the paragraph above to read:

> The `shouldUpdate` option can be useful to help avoid an unnecessary update, and therefore avoid an
unnecessary emission of `state$`. You might want to use this to avoid setting state to `null` when
it is already `null`. The `shouldUpdate` method gives you in it's first parameter the value of state
before any change has been made to it and the dependency you have, optionally, provided through
`combineLatestWith`. To avoid setting `null` to your state when it's already `null` you could call
`update` like below:

Do you think that is clearer?

Reply:
Yeah, it's more clear. I still think there's an issue, though. As written, the example will never send a request when its value is presently null. I think the sample would need to be adjusted to compare vs. some closed over variable to reflect what the paragraph is saying.

```typescript
const proposedValue = // ...
await myState.update(() => proposedvalue, { shouldUpdate: (currentValue) => currentValue !== null || proposedValue !== null });
```

Is that right?

Reply:
Spoke out-of-band about this. The missing context is **why** you'd want to set it to null. Since the code above categorically sets the value to null, as might occur in `clear()`, checking the current state is all that's needed. The sample I provided is also correct, but it introduces a `proposedValue`, as one might see when writing `save(value)`. Since the logic I'm working with always saves (there is no "delete" or "clear" logic), I'd assumed that context within the code.